core/src/main/java/org/apache/calcite/sql/dialect/MysqlSqlDialect.java

```diff
@@ -208,12 +212,52 @@ public boolean supportsAliasedValues() {
 
       unparseFloor(writer, call);
       break;
-
     default:
+      if (call.getOperator() == SqlStdOperatorTable.CONCAT) {
+        unparseConcat(writer, call);
+        break;
+      }
       super.unparseCall(writer, call, leftPrec, rightPrec);
     }
   }
 
+  /**
+   * Unparses String concatenation operator for Mysql, Mysql uses Concat as string concatenation
+   * operator when PIPES_AS_CONCAT is not enabled. CONCAT concatenation characters are used by
+   * default.
+   *
+   * @param writer writer
+   * @param call call
+   */
+  private void unparseConcat(SqlWriter writer, SqlCall call) {
+    List<SqlNode> expandOperandList = new ArrayList<>();
+    expandConcatOperands(call, expandOperandList);
```

Comment:
Why do we need to expand operands? `operand.unparse` can recursively handle nested case in nature.

Reply:
Expand operands for CONCAT, so we can generate CONCAT(`fname`, '_', '_', `lname`) instead of CONCAT(CONCAT(CONCAT(`fname`, '_'), '_'), `lname`)

Reply:
> Why do we need to expand operands? `operand.unparse` can recursively handle nested case in nature.

Expand operands for CONCAT, so we can generate CONCAT(fname, '', '', lname) instead of CONCAT(CONCAT(CONCAT(fname, ''), ''), lname)